docs/proposals/troubleshoot-running-pods.md

```diff
@@ -0,0 +1,568 @@
+that creates a Debug Container based on a provided `v1.Container` and attaches
+to its console. Users give Debug Containers a name (e.g. "debug" or "shell")
+which is used in subsequent interactions and is reported in `kubectl describe`.
+
+### Creating Debug Containers
+
+The process for creating a Debug Container is:
+
+1.  `kubectl` constructs a `v1.Container` based on command line flags and
+    `POST`s it to `/api/v1/namespaces/$NS/pods/$POD/debug`.
+1.  The API server performs admission control and proxies the connection to the
+    kubelet's `/podDebug/$NS/$POD/$DEBUG_CONTAINER_NAME` endpoint. `/podDebug`
+    is used because `/debug` is already used by the kubelet. `/podDebug` was
+    chosen to parallel existing endpoints like `/containerLogs`.
+1.  The kubelet instructs the Generic Runtime Manager (this feature is only
+    implemented for the CRI) to create a Debug Container.
+1.  The runtime manager uses the existing `startContainer()` method to create a
+    container in an existing pod. `startContainer()` has one modification for
+    Debug Containers: it creates a new runtime label (e.g. a docker label) that
+    identifies this container as a Debug Container.
+1.  The kubelet performs an attach operation and upgrades the client connection
+    to streaming.
+
+It is an error to attempt to create a Debug Container with the same name as a
+container that exists in the pod spec. There are no limits on the number of
+Debug Containers that can be created in a pod, but exceeding a pod's resource
+allocation may cause it to be evicted.
+
+### Debug Container Status
+
+The status of a *debug container* is reported in a new field in `v1.PodStatus`:
+
+```
+type PodStatus struct {
+        ...
+        DebugContainerStatuses []ContainerStatus
+}
+```
+
+This list is populated by the kubelet from
+`kubecontainer.PodStatus.ContainerStatuses` in the same way as regular and init
+container statuses. This is sent to the API server and displayed by `kubectl
+describe pod`.
+
+Status of running Debug Containers will always be reported in this list, but
+Debug Containers that exit will be removed from this list when they are garbage
+collected. Initially we will satisfy inspection and audit requirements by
+exempting Debug Containers from garbage collection, making a list of every debug
+container that has ever run in a pod. A more sophisticated solution will require
+a better understanding of reporting requirements. This will be required to
+graduate from alpha status.
+
+### Restarting and Reattaching Debug Containers
+
+Debug Containers will never be restarted automatically. It is possible to
+replace a Debug Container that has exited by re-using a Debug Container name. It
+is an error to attempt to replace a Debug Container that is still running.
+
+One way in which `kubectl debug` differs from `kubectl exec` is the ability to
+reattach to Debug Container if you've been disconnected. This is accomplished by
```

Comment:
Does the debug container exit when you disconnect from it? That would seem like generally desirable behavior to me, as it would guarantee that only debug containers that actively are being used stick around.